lib/logical_solver/presolve.hpp

```diff
@@ -0,0 +1,455 @@
+#include <string>
+#include <vector>
+#include <stack>
+#include <limits>
+#include <cmath>
+#include <iomanip>
+
+namespace logical_solver{
+
+  class Presolve
+  {
+  public:
+  // PUBLIC INSTANCE VARIABLES:
+  // problem
+  std::vector<std::vector<int>> problem_matrix_;
+  std::vector<double> lower_bounds_;
+  std::vector<double> upper_bounds_;
+
+  // Problem type
+  bool solve_MIP_;
+
+  // problem characteristics
+  int variables_count_;
+  int constraints_count_;
```

Comment:
If these are instantiated as mentioned in my other comment, they can be `const`.